### ISSUE TITLE
Write an autoformatter for Oil

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#681]
I don't have time to do this, but I think it belongs in the repo because of the relatively tight  coupling to the syntax tree.

Lots of good links in this thread:

https://news.ycombinator.com/item?id=22706242

But this one is simple  and gives the idea: the AST is compiled to an IR, and the IR is converted to formatted text of a given width.

https://blog.vjeux.com/2017/javascript/anatomy-of-a-javascript-pretty-printer.html

Note: there's also shfmt, but it handles bash and not Oil of course.

**Comment by andychu** [comment]
related to #478

**Comment by sanju50201** [comment]
can i work on this?

**Comment by andychu** [comment]
Yes, nobody is working on this right now!  There is a very related Zulip thread about this

https://oilshell.zulipchat.com/#narrow/stream/121539-oil-dev/topic/Translators.2C.20formatters.2C.20linters

We have kept the OSH -> YSH translator alive, and it can do an approximate translation

This is `osh --tool ysh-ify`

We could also have `osh --tool format`

(There is also the idea of exporting the AST so it be done outside the project, though that's a bit harder)

**Comment by Plastikov** [comment]
> can i work on this?

hello,

what is the status of your work, do you mind I join you to work on this togther?
